mod_wasm/modules/wasm/mod_wasm.c

```diff
@@ -437,105 +361,77 @@ static int content_handler(request_rec *r)
  */
 static void register_hooks(apr_pool_t *p)
 {
-    ap_hook_post_config(post_config_hook, NULL, NULL, APR_HOOK_MIDDLE);
     ap_hook_handler(content_handler, NULL, NULL, APR_HOOK_MIDDLE);
 }
 
 
-#define WASM_DIRECTIVE_WASMROOT   "WasmRoot"
-#define WASM_DIRECTIVE_WASMMODULE "WasmModule"
-#define WASM_DIRECTIVE_WASMARG    "WasmArg"
-#define WASM_DIRECTIVE_WASMENV    "WasmEnv"
-#define WASM_DIRECTIVE_WASMDIR    "WasmDir"
-#define WASM_DIRECTIVE_WASMMAPDIR "WasmMapDir"
-#define WASM_DIRECTIVE_ENABLECGI  "WasmEnableCGI"
-
-
-static const char *wasm_directive_WasmRoot(cmd_parms *cmd, void *mconfig, const char *word1)
-{
-    x_cfg *cfg = (x_cfg *) mconfig;
-    wasm_config_set_root(word1);
-    return NULL;
-}
-
+#define WASM_DIRECTIVE_WASMMODULE     "WasmModule"
+#define WASM_DIRECTIVE_WASMARG        "WasmArg"
+#define WASM_DIRECTIVE_WASMENV        "WasmEnv"
+#define WASM_DIRECTIVE_WASMDIR        "WasmDir"
+#define WASM_DIRECTIVE_WASMMAPDIR     "WasmMapDir"
+#define WASM_DIRECTIVE_WASMENABLECGI  "WasmEnableCGI"
 
 static const char *wasm_directive_WasmModule(cmd_parms *cmd, void *mconfig, const char *word1)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    wasm_config_set_module(word1);
+    wasm_module_load(word1);    // Wasm module is loaded and cached
+    wasm_config_set_module(cfg->loc, word1);    // Wasm config is implictly created for the current location and using the loaded module
     return NULL;
 }
 
 
 static const char *wasm_directive_WasmArg(cmd_parms *cmd, void *mconfig, const char *word1)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    if (cfg->configArgCount == CONFIG_DEFINED_ARGS_MAX) {
-      return apr_psprintf(cmd->pool, "WasmArg limit reached in the httpd configuration (maximum is %d)", CONFIG_DEFINED_ARGS_MAX);
-    }
-    configArg *arg = malloc(sizeof(configArg));
-    arg->arg = apr_pstrdup(cmd->pool, word1);
-    cfg->configArgs[cfg->configArgCount] = arg;
-    cfg->configArgCount++;
+    wasm_config_arg_add(cfg->loc, word1);
```

Comment:
Shouldn't we check the return code here?

Reply:
I agree